### DISCUSSION TITLE
Having some trouble understanding File Storage

### DISCUSSION THREAD

**Discussion by dev-ref** [pocketbase/pocketbase#1638] [Q&A]
I'm working on a project that involves recording a 30 second clip of audio from the user, but I'm struggling to set up the function to call on a "Submit" button. I'd like it to go to the 'messages' collection, also preferably with the username as the filename. I tried using [the docs](https://pocketbase.io/docs/files-handling) but I struggled with that for some time. I think the area where I'm getting turned around is understanding 'collections' specifically, though I do know that mediaRecorder hasn't been the easiest either.
I'll append the current code here, though it's probably unnecessary and it doesn't contain my attempts thus far. I don't particularly want somebody to write the code for me, only to help me understand the sequence of events I need to take.

```

<script lang="ts">
    import { onMount } from 'svelte'
    import PocketBase from 'pocketbase';
    let media = [];
    let mediaRecorder = null;
    let timeoutId;
    onMount(async () => {
      const stream = await navigator.mediaDevices.getUserMedia({ audio: true });
      mediaRecorder = new MediaRecorder(stream);
      mediaRecorder.ondataavailable = (e) => media.push(e.data)
      mediaRecorder.onstop = function(){
        const audio = document.querySelector('audio');
        const blob = new Blob(media, {'type' : 'audio/ogg; codecs=opus' });
        media = [];
        audio.src = window.URL.createObjectURL(blob);
        const file = new File([blob], "audio.ogg", {type: "audio/ogg"});
        const formData = new FormData();
        formData.append("audio", file);
        fetch('insert_URL:PORT', {
          method: 'POST',
          body: formData
        })
        .then(response => response.json())
        .then(data => {
        })
        .catch(error => {
          console.error('Error:', error);
        });
      }
    })
    function startRecording(){
      mediaRecorder.start();
      timeoutId = setTimeout(() => {
        mediaRecorder.stop();
      }, 30100);
    }
    function stopRecording() {
      clearTimeout(timeoutId);
      mediaRecorder.stop();
    }
    
  </script>
  
    <section>
      <audio controls />
      <button on:click={startRecording}>Record</button> 
      <button on:click={stopRecording}>Stop Recording</button>
      <button on:click-{submit}>Submit</button>
  </section>
```

**Comment by ganigeorgiev** [comment]
At first glance the above sequence seems correct.

What is the error you are getting?

Since you are using `fetch` directly, make sure that the url it is pointing to is correct, eg.:
```js
fetch('http://127.0.0.1:8090/api/collections/YOUR_COLLECTION_NAME_OR_ID/records', { ... });
```

Alternatively, you can also use the SDK since you already import it:

```js
const pb = new PocketBase("http://127.0.0.1:8090")

...

// optionally authenticate if your collection API rule requires it
// await pb.collection("users").authWithPassword("test@example.com", "1234567890")

const record = await pb.collection("yourCollection").create(formData)
```

**Reply by ganigeorgiev** [reply]
Additionally, to help you with the debugging in case of a server-side error, you can also start PocketBase with the `--debug` flag and inspect the console output and/or check the request logs in the "Admin UI > Logs".

**Reply by dev-ref** [reply]
Thanks for the response! I was able to write the file to the appropriate collections record. The error I was having was related to an improperly configured 'user' field, as I was trying to make sure that the user recording the message had their username attached to the filename. Currently I still don't have that functionality, but a bigger issue I'm having now is retrieving the file from the database so I can listen to it at a later time. I attempted to create a "Recordings" function to allow a list to be visible to users who fulfill a certain role, which currently does not work and is not supplying me any errors.
```
<script>
    import { onMount } from 'svelte';
    import { pb } from './pocketbase';

    let recordings = [];
  
    onMount(async () => {
      // Get all recordings from the 'messages' collection
      const resultList = await pb.collection('messages').getList(1, 50, {
          sort: 'created',
          expand: 'user',
      });
      recordings = resultList.items;
    });
  
    function playRecording(id) {
      // Get the selected recording from the 'messages' collection
      pb.collection('messages').getOne(id)
      .then((audioFile) => {
          const blob = new Blob([audioFile.audio], { type: 'audio/ogg' });
          const url = URL.createObjectURL(blob);
          const audio = new Audio(url);
          audio.play();
      })
      .catch((error) => {
          console.error('Error:', error);
      });
    }
  </script>
  
  <div>
    <h3>Recordings:</h3>
    {#each recordings as recording (recording.id)}
      <div>
        <p>{recording.expand?.user?.username}</p>
        <button on:click={() => playRecording(recording.id)}>Play</button>
      </div>
    {/each}
  </div>
```
I'm also attaching a screenshot of the 'messages' collection here.
![messages-collection](https://user-images.githubusercontent.com/121682054/213419647-81978728-07b7-4e48-89b2-41fbe7c5469f.PNG)

**Reply by ganigeorgiev** [reply]
You don't need to create a blob. You can access directly the public file url  https://pocketbase.io/docs/files-handling/#file-url. You can also use the `getFileUrl` helper:

```js
.then((audioFile) => {
  const audio = new Audio(pb.getFileUrl(audioFile, audioFile.yourFileField));
  audio.play(); // may need to be wrapped inside an event?
})
```

**Reply by ganigeorgiev** [reply]
Also make sure to check https://developer.mozilla.org/en-US/docs/Web/API/HTMLAudioElement/Audio#determining_when_playback_can_begin.
